[Feat] Add dynamic data and caching to data manager (#398)
mckinsey/vizro: vizro-core/docs/pages/user-guides/dynamic-data.md

# Dynamic data

> A dynamic data source is a Python function that returns a pandas DataFrame. This function is executed when the dashboard is initially started and _can be executed again while the dashboard is running_. This makes it possible to refresh the data shown in your dashboard without restarting the dashboard itself. If you do not this functionality then you should use [static data](static-data.md) instead.

Reviewer: ```suggestion
A dynamic data source is a Python function that returns a pandas DataFrame. This function is executed when the dashboard is initially started and _can be executed again while the dashboard is running_. This makes it possible to refresh the data shown in your dashboard without restarting the dashboard itself. If you do not require this functionality then you should use [static data](static-data.md) instead.
```

Unlike static data, dynamic data cannot be supplied directly into the `data_frame` argument of a `figure`. Instead, it must first be added to the Data Manager and then referred to by name.

> !!! example "Dynamic data"

Author: Just as a quick thing for now I will do something where it just plots the iris dataset but with different random points selected or something like that. And then in the future we can make this a more interesting example like the weather where we call some external API.

> !!! example "Dynamic data"

Author: Done in [acb546a](https://github.com/mckinsey/vizro/pull/398/commits/acb546a8a76658ab09b1c9e1b6c3b7559dfc694e) - see what you think.

@maxschulz-COL you don't know of any reason why we wouldn't be able to host `iris.csv` on RTD and download it as done here?

> !!! example "Dynamic data"

Reviewer: No, but I also haven't tried it. But I like the new approach. The only thing I am unsure about is the following:

Should the code example really have `pd.read_csv` (I know it is to make it clear you are fetching from a file) or actually `px.data.iris()`, ie something that will always work?

I kinda prefer the former, but with a message similar to the one we already have, ie that here we kinda simulate the dynamic case

> !!! example "Dynamic data"

Author: Yeah, I had this done with `px.data.iris()` originally and then decided that wasn't clear enough because it's too far removed from a "real" working example where the data can change. So better to use `pd.read_csv` even though that means downloading a file to get the example working.

The download works ok on the automatically built docs for this PR so presumably it will also work with the released docs 🤞

## Configure cache

By default, every time the dashboard is refreshed a dynamic data function will be executed again. This means that your dashboard will always show the very latest data. In fact, if there are multiple graphs on the same page using the same dynamic data source then the loading function will be executed _multiple_ times, once for each graph on the page. Hence, if loading your data is a slow operation, your dashboard performance may suffer.

The Vizro Data Manager has a caching mechanism to help solve this. Vizro's cache uses [Flask-Caching](https://flask-caching.readthedocs.io/en/latest/), which supports a number of possible cache backends and [configuration options](https://flask-caching.readthedocs.io/en/latest/#configuring-flask-caching). By default, the cache is turned off.

In a development environment the easiest way to enable caching is to use a [simple memory cache](https://cachelib.readthedocs.io/en/stable/simple/) with the default configuration options:

```py title="Simple cache with default timeout of 5 minutes"
from flask_caching import Cache

data_manager.cache = Cache(config={"CACHE_TYPE": "SimpleCache"})
data_manager["iris"] = load_iris_data
```

> By default, dynamic data is cached in the Data Manager for 5 minutes. A refresh of the dashboard within this time interval will fetch the pandas DataFrame from the cache and not re-run the data loading function. Once the cache timeout period has elapsed, the next refresh of the dashboard will re-execute the dynamic data loading function. The resulting pandas DataFrame will again be put into the cache and not expire until another 5 minutes has elapsed.

Author: Agreed on lowercase `data manager` with @stichbury and implemented in [6d8eaab](https://github.com/mckinsey/vizro/pull/398/commits/6d8eaab881272cbd6c7d5881c2372f03621a2d4b).

If you would like to alter some options, such as the default cache timeout, then you can specify a different cache configuration:

```py title="Simple cache with timeout set to 10 minutes"

> data_manager.cache = Cache(config={"CACHE_TYPE": "SimpleCache", "CACHE_DEFAULT_TIMEOUT": 600})

Reviewer: Is it possible to set the cache to just refresh when the source has been modified, e.g. when it detects a file modification since the last load?

I am thinking here of the model interactions case - when the kedro run has finish, then the charts should be reloaded

Author: Hmm, it's probably possible to update cache automatically when a file changes but not very easy.

What will be easy though, and coming in the near future in a different PR, is to have an action that forcibly updates data without waiting for it to expire. So you could have something that does:
1. action to trigger kedro run
2. action to refresh data

```

!!! warning

    Simple cache exists purely for single-process development purposes and is not intended to be used in production. If you deploy with multiple workers, [for example with gunicorn](run.md/#gunicorn), then you should use a production-ready cache backend. All of Flask-Caching's [built-in backends](https://flask-caching.readthedocs.io/en/latest/#built-in-cache-backends) other than `SimpleCache` are suitable for production. In particular, you might like to use [`FileSystemCache`](https://cachelib.readthedocs.io/en/stable/file/) or [`RedisCache`](https://cachelib.readthedocs.io/en/stable/redis/):

    ```py title="Production-ready caches"
    # Store cached data in CACHE_DIR
    data_manager.cache = Cache(config={"CACHE_TYPE": "FileSystemCache", "CACHE_DIR": "cache"})

    # Use Redis key-value store
    data_manager.cache = Cache(config={"CACHE_TYPE": "RedisCache", "CACHE_REDIS_HOST": "localhost", "CACHE_REDIS_PORT": 6379})
    ```

> ### Configure timeouts

Author: This is meant to be a subsection because it only makes sense to configure timeouts if you have already configured the cache. Let me change the title to make that clearer though.

    ```py title="Production-ready caches"
    # Store cached data in CACHE_DIR
    data_manager.cache = Cache(config={"CACHE_TYPE": "FileSystemCache", "CACHE_DIR": "cache"})

    # Use Redis key-value store
    data_manager.cache = Cache(config={"CACHE_TYPE": "RedisCache", "CACHE_REDIS_HOST": "localhost", "CACHE_REDIS_PORT": 6379})
    ```

Note that when a production-ready cache backend is used, the cache is persisted beyond the Vizro process and is not cleared by restarting your server. If you wish to clear the cache then you must do so manually, e.g. if you use `FileSystemCache` then you would delete your `cache` directory. Persisting the cache can also be useful for development purposes when handling data that takes a long time to load: even if you do not need the data to refresh while your dashboard is running, it can speed up your development loop to use dynamic data with a cache that is persisted between repeated runs of Vizro.

### Configure timeouts

> You can change the timeout of the cache independently for each dynamic data source in the Data Manager using the `timeout` setting (measured in seconds). A `timeout` of 0 indicates that the cache does not expire. This is effectively the same as using [static data](static-data.md).

Author: I thought before about adding an option for "always expires" `timeout` data and decided not to bother. For now the solution is just to set a low timeout (the lowest possible is `timeout = 1`). Given that we have to handle network requests, which takes a while, and don't really deal with live-streamed data pushed through websocket or similar anyway, I think 1 second timeout basically counts good enough as "always expires" for our purposes.

(To me, ideally `timeout=0` sound like it should mean "always expires", and `timeout=inf` would mean "never expires". I considered implementing this but decided changing the meaning of `timeout=0` from what flask-caching does was too confusing, even though I think they've got the meaning of `timeout=0` the wrong way round...)

Does that make sense?

> You can change the timeout of the cache independently for each dynamic data source in the Data Manager using the `timeout` setting (measured in seconds). A `timeout` of 0 indicates that the cache does not expire. This is effectively the same as using [static data](static-data.md).

Author: Hmm maybe yes, good idea - let me see how easy it is to add!

> You can change the timeout of the cache independently for each dynamic data source in the Data Manager using the `timeout` setting (measured in seconds). A `timeout` of 0 indicates that the cache does not expire. This is effectively the same as using [static data](static-data.md).

Reviewer 2: I also thought about `timeout=0 # always expires` and `timeout=None # (or False) - never expires` option and it perfectly makes sense. However, I don't know how wise it is to change the behaviour for `timeout` comparing on how it behaves in the flask_cacking lib. 

Maybe we should stick with `timeout=0 # never expires` and then investigate and implement the option for `always expires` afterwards in a separate PR. Otherwise we need to implement it right now or to break it in the future?

> You can change the timeout of the cache independently for each dynamic data source in the Data Manager using the `timeout` setting (measured in seconds). A `timeout` of 0 indicates that the cache does not expire. This is effectively the same as using [static data](static-data.md).

Author: Yeah, I had exactly the same thoughts as you and decided it's also kind of confusing to have `timeout = 0` meaning the opposite of `timeout = False`, given that `False` is almost the same as `0` 😬 So probably `timeout = data_manager.ALWAYS` where `ALWAYS` is some sentinel value would be best here to be unambiguous.

Let's leave with the flask-caching convention that `timeout = 0` means "never expires" and then see whether people actually want an "alway expires" setting or if just setting a low timeout like `timeout = 1` is sufficient for user needs. If it's not then in future we can add `timeout = ALWAYS` without breaking anything.

# Dynamic data

A dynamic data source is a Python function that returns a pandas DataFrame. This function is executed when the dashboard is initially started and _can be executed again while the dashboard is running_. This makes it possible to refresh the data shown in your dashboard without restarting the dashboard itself. If you do not require this functionality then you should use [static data](static-data.md) instead.

Unlike static data, dynamic data cannot be supplied directly into the `data_frame` argument of a `figure`. Instead, it must first be added to the Data Manager and then referenced by name.

> The below toy example demonstrates how dynamic data is updated every time the page is refreshed. When you run the code and refresh the page the data shown will change every time because the function `load_iris_data` is re-run. The example uses the Iris dataset saved to a file `iris.csv` in the same directory as `app.py`. This data can be generated using `px.data.iris()` or [downloaded](../../assets/user_guides/data/iris.csv).

Reviewer 3: ```suggestion
The example below shows how dynamic data is updated every time the page is refreshed. When you run the code and refresh the page the data shown will change every time because the function `load_iris_data` is re-run. The example uses the Iris dataset saved to a file `iris.csv` in the same directory as `app.py`. This data can be generated using `px.data.iris()` or [downloaded](../../assets/user_guides/data/iris.csv).
```

Reviewer 3: > The below toy example demonstrates how dynamic data is updated every time the page is refreshed. When you run the code and refresh the page the data shown will change every time because the function `load_iris_data` is re-run. The example uses the Iris dataset saved to a file `iris.csv` in the same directory as `app.py`. This data can be generated using `px.data.iris()` or [downloaded](../../assets/user_guides/data/iris.csv).

I think we need to rephrase this a bit as the suggestion is that if you refresh the page you change the data. In most realtime systems, I'd expect the opposite (the data changes, the page refreshes with the update) but I guess we aren't claiming that here. So what is happening is that the page is updated the data is fetched again, right? Could we try this suggestion...?

Reviewer 3: ```suggestion
The example below shows how data is fetched dynamically every time the page is refreshed. When you run the code and refresh the page the data is fetched by the function `load_iris_data` is re-run. The example uses the Iris dataset saved to a file `iris.csv` in the same directory as `app.py`. This data can be generated using `px.data.iris()` or [downloaded](../../assets/user_guides/data/iris.csv).

```

Reviewer 3: I know it's subtle but it removes (to my mind) any chance people get confused that a change in the data causes an update of the UI. Which is not what is happening AFAICT.

Author: You are absolutely right, thank you for the suggestion! I will make the change.

            ],
            controls=[vm.Filter(column="species")],
        )

        dashboard = vm.Dashboard(pages=[page])

        Vizro().build(dashboard).run()
        ```

        1. `iris` is a pandas DataFrame created by reading from the CSV file `iris.csv`.
        2. To demonstrate that dynamic data can change when the page is refreshed, select 30 points at random. This simulates what would happen if your file `iris.csv` were constantly changing.
        3. To use `load_iris_data` as dynamic data it must be added to the Data Manager. You should **not** actually call the function as `load_iris_data()`; doing so would result in static data that cannot be reloaded.
        4. Dynamic data is referenced by the name of the data source `"iris"`.

    === "Result"
        [![DynamicData]][DynamicData]

    [DynamicData]: ../../assets/user_guides/data/dynamic_data.gif

Since dynamic data sources must always be added to the Data Manager and referenced by name, they may be used in YAML configuration [exactly the same way as for static data sources](static-data.md#reference-by-name).

## Configure cache

> By default, every time the dashboard is refreshed a dynamic data function will be executed again. This means that your dashboard will always show the very latest data. In fact, if there are multiple graphs on the same page using the same dynamic data source then the loading function will be executed _multiple_ times, once for each graph on the page. Hence, if loading your data is a slow operation, your dashboard performance may suffer.

Reviewer 3: ```suggestion
By default, each time the dashboard is refreshed a dynamic data function executes again. In fact, if there are multiple graphs on the same page using the same dynamic data source then the loading function executes _multiple_ times, once for each graph on the page. Hence, if loading your data is a slow operation, your dashboard performance may suffer.
```

Reviewer 3: I removed "This means that your dashboard will always show the very latest data." because I don't think this is true -- it's only the data that was current at the time the dashboard refreshed, and it may update after that (unless the update to the data triggers another refresh?).

Author: You are right 👍 it is indeed only the data that's current at the time of refresh - when the data is updated, it doesn't get automatically pushed to the frontend.
